tests/test_emulation.py

```diff
@@ -0,0 +1,238 @@
+        """It should detect touch when applying viewport with touches."""
+        page = isolated_page
+        await page.goto(server / 'detect-touch.html')
+        assert await page.evaluate("document.body.textContent.trim()") == 'NO'
+        await page.setViewport(iPhone['viewport'])
+        await page.goto(server / 'detect-touch.html')
+        assert await page.evaluate("document.body.textContent.trim()") == 'YES'
+
+    @chrome_only
+    @sync
+    async def test_is_detecteable_by_modernizr(self, server, isolated_page):
+        """Verify the emulation should be detectable by Modernizr JS lib."""
+        page = isolated_page
+        await page.setViewport({"width": 800, "height": 600, "hasTouch": True})
+        await page.addScriptTag(url=f"{server / 'modernizr.js'}")
+        assert await page.evaluate("Modernizr.touchevents") == True
+
+    @chrome_only
+    @sync
+    async def test_supports_landscape_emulation(self, server, isolated_page):
+        """It should support landscape emulation."""
+        page = isolated_page
+        await page.goto(server / 'mobile.html')
+        assert await page.evaluate("screen.orientation.type") == 'portrait-primary'
+        await page.setViewport(iPhoneLandscape['viewport'])
+        assert await page.evaluate("screen.orientation.type") == 'landscape-primary'
+        await page.setViewport({"width": 100, "height": 100})
+        assert await page.evaluate("screen.orientation.type") == 'portrait-primary'
+
+
+class TestPageEmulation:
+    @sync
+    async def test_page_emulates_mobile_devices(self, server, isolated_page):
+        """Verify the page emulation works with emulating mobile devices."""
+        page = isolated_page
+        await page.goto(server / 'mobile.html')
+        await page.emulate(viewport=iPhone['viewport'], userAgent=iPhone['userAgent'])
+        assert await page.evaluate("window.innerWidth") == 375
+        assert 'iPhone' in await page.evaluate("navigator.userAgent")
+
+    @sync
+    async def test_supports_clicking(self, server, isolated_page):
+        """Verify the emulation should support clicking."""
+        page = isolated_page
+        await page.emulate(viewport=iPhone['viewport'], userAgent=iPhone['userAgent'])
+        await page.goto(server / 'input/button.html')
+        button = await page.J('button')
+        await page.evaluate("(button) => (button.style.marginTop = '200px')", button)
+        await button.click()
+        assert await page.evaluate("result") == 'Clicked'
+
+
+# deprecated
+class TestEmulationMedia:
+    """
+    emulateMedia is deprecated in favour of emulateMediaType but we
+    don't want to remove it from Puppeteer just yet. We can't check
+    that emulateMedia === emulateMediaType because when running tests
+    with COVERAGE=1 the methods get rewritten. So instead we
+    duplicate the tests for emulateMediaType and ensure they pass
+    when calling the deprecated emulateMedia method.
```

Comment:
Could you elaborate on what you mean here? Specifically:

> We can't check that emulateMedia === emulateMediaType because when running tests with COVERAGE=1 the methods get rewritten

Reply:
Simplified expalanation

Reply:
Is this really necessary for coverage though?

Reply:
Agree.
Still I left one test as a part of TestEmulateMediaType to maintain coverage